### ISSUE TITLE
Do we need to flatten the parameters in PyTorchCost

### ISSUE THREAD

**Issue by nicola-bastianello** [team-decent/decent-bench#291]
currently, `PyTorchCost` always flattens the internal parameters before returning. since `Array` can have whatever shape, I'm wondering if we can skip the flattening and just pass multi-dimensional tensors around

**Comment by Simpag** [comment]
This is something that should be tested. I did this so that linear algebra operations would work as expected during algorithm execution. I believe that element-wise operations, +, -, *, /, would all work as expected but matrix multiplication might be strange. I am not sure if a @ b will be the same value if a multi-dimensional tensor is used. When I originally designed the pytorch cost I ran into some issue but I cant remember what it was. 

But it would make it a bit more efficient to not flatten the tensor all of the time, however its not a huge deal as they are only views and not a full reshape + copy, except possibly for the last torch.cat call on all the vectors but I am not sure. 

I am not sure if the setting of model parameters will be much faster if we dont flatten the return as I believe we still need to iterate through the list of parameters and set them individually, not flattening would remove one .view() call which reshapes the tensor without copying it but I think this is a very lightweight call. 

Some issues may arise when users create buffers and expect the optimization variable to be a 1D vector. But this could be solved by changing the pytorch cost shape parameter to be the multidimensional shape and users create buffers along the lines of `buffer = iop.zeros((len(neighbors), *agent.cost.shape))`

**Comment by nicola-bastianello** [comment]
> This is something that should be tested. I did this so that linear algebra operations would work as expected during algorithm execution. I believe that element-wise operations, +, -, *, /, would all work as expected but matrix multiplication might be strange. I am not sure if a @ b will be the same value if a multi-dimensional tensor is used. When I originally designed the pytorch cost I ran into some issue but I cant remember what it was.

I think `torch.matmul` is capable of handling a @ b even with tensors, provided the dimensions make sense. but yes there might be issues as in some places we assume to work with a vector

> But it would make it a bit more efficient to not flatten the tensor all of the time, however its not a huge deal as they are only views and not a full reshape + copy, except possibly for the last torch.cat call on all the vectors but I am not sure.
> 
> I am not sure if the setting of model parameters will be much faster if we dont flatten the return as I believe we still need to iterate through the list of parameters and set them individually, not flattening would remove one .view() call which reshapes the tensor without copying it but I think this is a very lightweight call.

I see, I agree that setting the parameters is likely the more time-consuming operation

> Some issues may arise when users create buffers and expect the optimization variable to be a 1D vector. But this could be solved by changing the pytorch cost shape parameter to be the multidimensional shape and users create buffers along the lines of `buffer = iop.zeros((len(neighbors), *agent.cost.shape))`

indeed, and I think that even if we keep flattening we could still make this change to using multidimensional shape, so that later it is easier to test if unflattened tensors still work

**Comment by Simpag** [comment]
> > This is something that should be tested. I did this so that linear algebra operations would work as expected during algorithm execution. I believe that element-wise operations, +, -, *, /, would all work as expected but matrix multiplication might be strange. I am not sure if a @ b will be the same value if a multi-dimensional tensor is used. When I originally designed the pytorch cost I ran into some issue but I cant remember what it was.
> 
> I think `torch.matmul` is capable of handling a @ b even with tensors, provided the dimensions make sense. but yes there might be issues as in some places we assume to work with a vector
> 

All of the frameworks are able to work with multi-dimensions matrix multiplications. Only issue is if a user expects for example a dot product when doing `iop.transpose(x) @ x` but when working with multi dimensional tensors this wont be a scalar value. Furthermore, the transpose operator only applies to the last two dimensions so a transpose of a multi-dimensional tensor is not as intuitive. However, since algorithms are implemented from an agent perspective matrix multiplication might not come up. 

Regarding all of this flattening. We should do some profiling of the codebase before we make any changes. From my previous profiling the flattening + unflattening is a small portion of the time spent as I tried to optimize it as much as possible using a combination of pytorch capabilities. But this may scale depending on network size and I profiled this using a rather small network. Removing the flattening + unflattening should be a small change so it should be possible to do some empirical testing when I push my MNIST + NIM examples (while making sure that the results remain the same).

On a side note, the performance of algorithms would greatly increase if we supported more general interpretations of algorithms where agent states are encoded in matrices and algorithms perform matrix operations. While most papers present matrix interpretations of their algorithms, the algorithm presentation tends to be from an agent perspective so it might not make a whole lot of sense. But then again, each agent would need to perform their own gradient calculations and this is the most expensive operation so the performance gain might not be large. Lets not do this.

**Comment by nicola-bastianello** [comment]
> All of the frameworks are able to work with multi-dimensions matrix multiplications. Only issue is if a user expects for example a dot product when doing `iop.transpose(x) @ x` but when working with multi dimensional tensors this wont be a scalar value. Furthermore, the transpose operator only applies to the last two dimensions so a transpose of a multi-dimensional tensor is not as intuitive. However, since algorithms are implemented from an agent perspective matrix multiplication might not come up.

I see, indeed it might give unexpected results. but I agree that matrix multiplication is unlikely to come up with the current set-up. and as long as we expose `iop.norm` users hopefully won't need `iop.transpose(x) @ x` or similar

> Regarding all of this flattening. We should do some profiling of the codebase before we make any changes. From my previous profiling the flattening + unflattening is a small portion of the time spent as I tried to optimize it as much as possible using a combination of pytorch capabilities. But this may scale depending on network size and I profiled this using a rather small network. Removing the flattening + unflattening should be a small change so it should be possible to do some empirical testing when I push my MNIST + NIM examples (while making sure that the results remain the same).

yes, let's profile this on bigger benchmarks and then decide

> On a side note, the performance of algorithms would greatly increase if we supported more general interpretations of algorithms where agent states are encoded in matrices and algorithms perform matrix operations. While most papers present matrix interpretations of their algorithms, the algorithm presentation tends to be from an agent perspective so it might not make a whole lot of sense. But then again, each agent would need to perform their own gradient calculations and this is the most expensive operation so the performance gain might not be large. Lets not do this.

I've thought a lot about this when I started the project. it is true that most (actually, all) algorithms can be written in matrix form. however, asynchronous algorithms are more difficult to write in matrix form (one needs to introduce "activation" matrices that are one in the elements corresponding to active agents, zero otherwise). so in my opinion it is actually much more involved to code an asynchronous algorithm from a matrix perspective. plus the creation of these "activation matrices" might actually add a good amount of overhead

also, as you say I don't foresee switching to a matrix form to give that much performance gain

**Comment by nicola-bastianello** [comment]
I've looked a bit into this, and it appears that avoiding flattening might not be possible since pytorch stores parameters as a dict of tensors. but these tensors might have different sizes, so stacking into a non-flattened array is not possible in general.

what we could do, though, is try to make the internals of PyTorchCost more efficient, trying to perform a single flattening operation. copilot had some suggestions so this might be possible
